### PR TITLE
 stat/distuv: add logistic distribution

### LINKED ISSUE #1744: Add Logistic distribution
### What are you trying to do?

I am trying to use [Logistic](https://en.wikipedia.org/wiki/Logistic_distribution) CDF.

### How does Gonum not allow you to achieve your goal?

Gonum does have a number of distributions but Logistic distribution is not one of them.  
It is used in logistic regression and also particularly useful for rating calculations (e.g. Elo rating).

### What version of Go and Gonum are you using?

go version go1.17.2 linux/amd64
gonum v0.9.3

### Is this feature absent from the current master?
Yes

### Are you able to help contribute the feature?

I suppose I can write an implementation but I am not familiar with gonum internals to do so efficiently.

### PR DESCRIPTION
**Pull request by gudvinr**
This PR adds [Logistic distribution](https://en.wikipedia.org/wiki/Logistic_distribution) to distuv package.

Tests cover edge cases for functions of this distribution.

### CODE REVIEW

File: stat/distuv/logistic.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
+package distuv
+
+import (
+	"math"
+)
+
+// Logistic implements the Logistic distribution. It is two-parameter distribution.
```
Comment (romanwerpachowski) [inline stat/distuv/logistic.go:7]:
```suggestion
// Logistic implements the Logistic distribution, a two-parameter distribution with support on the real axis.
```

File: stat/distuv/logistic_test.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
+package distuv
+
+import (
+	"math"
+	"testing"
+)
+
```
Comment (romanwerpachowski) [inline stat/distuv/logistic_test.go:7]: Please add `t.Parallel()` at the start of each test.

File: stat/distuv/logistic_test.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+func TestLogisticParameters(t *testing.T) {
+	l := Logistic{Mu: 0, S: 0}
+
+	if l.NumParameters() != 2 {
+		t.Fail()
```
Comment (romanwerpachowski) [inline stat/distuv/logistic_test.go:12]: Please use `t.Errorf(...)`, like for example in https://github.com/gonum/gonum/blob/master/stat/distuv/beta_test.go

File: stat/distuv/logistic_test.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+	l = Logistic{Mu: 0, S: 1}
+
+	if !math.IsInf(l.Quantile(0), -1) {
+		t.Fail()
+	}
+
+	if !math.IsInf(l.Quantile(1), 1) {
+		t.Fail()
+	}
```
Comment (romanwerpachowski) [inline stat/distuv/logistic_test.go:114]: No tests for finite results?

File: stat/distuv/logistic_test.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	if l.StdDev() != 1 {
+		t.Fail()
+	}
+
+	if l.Variance() != 1 {
+		t.Fail()
+	}
+}
+
+func TestLogisticCDF(t *testing.T) {
```
Comment (romanwerpachowski) [inline stat/distuv/logistic_test.go:49]: Would be nice to see some tests for PDF and CDF for some general values of parameters and argument, comparing the result with e.g. value returned by a reference implementation (scipy? Mathematica? Matlab?).

**Review by romanwerpachowski — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution! Implementation looks good (some small nitpicks), but the tests need more alignment with the existing codebase (see comments).

**Comment by kortschak** [comment]
Please note the test failures. These are code formatting and so. You can run the tests locally by running the scripts in .github/workflows that are listed in the CI config [here](https://github.com/gonum/gonum/blob/295bfca71ff005698ffd1d90261a2a8cf9183eb3/.github/workflows/ci.yml#L81-L88).

Also note the [format convention for commit messages](https://github.com/gonum/gonum/blob/295bfca71ff005698ffd1d90261a2a8cf9183eb3/.github/PULL_REQUEST_TEMPLATE#L12-L15) (we'll fix this on merge); the current PR title matches this convention.

**Comment by gudvinr** [comment]
> Please note the test failures. These are code formatting and so on.

That should do it

File: stat/distuv/logistic_test.go (hunk 5 above)
Comment (gudvinr) [reply stat/distuv/logistic_test.go:70]: Added couple of test values from WolframAlpha

**Comment by gudvinr** [comment]
@romanwerpachowski tests can be reviewed again

@kortschak rebased with correct message. Although I'd prefer everything to be squashed to single commit before merge

**Comment by kortschak** [comment]
Yes, we'll squash this when it's all ready. Also we will ask you to send another PR adding yourself to the AUTHORS and CONTRIBUTORS files with the commit message "A+C: add Maxim S" (if you're happy to have your full name in the files, please replace the S). This doesn't need to happen immediately, but we will need it to be ready before we merge this.

**Comment by gudvinr** [comment]
> Also we will ask you to send another PR adding yourself to the AUTHORS and CONTRIBUTORS files

Sure thing: #1746

**Review by romanwerpachowski — APPROVED** [review]
Thank you!

File: stat/distuv/logistic.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+)
+
+// Logistic implements the Logistic distribution, a two-parameter distribution with support on the real axis.
+// Its cumulative distribution function is the logistic function.
+//
+// General form of probability density fuction for Logistic distribution is
+//  E(x) / (s * (1 + E(x))^2)
+//  where E(x) = exp(-(x-μ)/s)
+//
+// For more information, see https://en.wikipedia.org/wiki/Logistic_distribution
```
Comment (kortschak) [inline stat/distuv/logistic.go:18]: Final dot.

File: stat/distuv/logistic.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+// Survival returns the survival function (complementary CDF) at x.
+func (l Logistic) Survival(x float64) float64 {
+	return 1 - l.CDF(x)
+}
+
+// Variance returns the variance of the probability distribution.
+func (l Logistic) Variance() float64 {
+	return math.Pow(l.StdDev(), 2)
```
Comment (kortschak) [inline stat/distuv/logistic.go:97]: This would be cheaper as `l.S*l.S*math.Pi*math.Pi/3`.

File: stat/distuv/logistic.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+// Mean returns the mean of the probability distribution.
+func (l Logistic) Mean() float64 {
+	return l.Mu
+}
+
+// Mode returns the mode of the distribution.
+//
+// It is same as Mean for Logistic distribution
```
Comment (kortschak) [inline stat/distuv/logistic.go:47]: Final dot.

File: stat/distuv/logistic.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+func (l Logistic) Mode() float64 {
+	return l.Mu
+}
+
+// Median returns the median of the distribution.
+//
+// It is same as Mean for Logistic distribution
```
Comment (kortschak) [inline stat/distuv/logistic.go:54]: Dot.

File: stat/distuv/logistic.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+func (l Logistic) Median() float64 {
+	return l.Mu
+}
+
+// NumParameters returns the number of parameters in the distribution.
+//
+// Always returns 2
```
Comment (kortschak) [inline stat/distuv/logistic.go:61]: Dot.

File: stat/distuv/logistic.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+// Quantile returns the inverse of the cumulative distribution function.
+func (l Logistic) Quantile(p float64) float64 {
+	return l.Mu + l.S*math.Log(p/(1-p))
+}
+
+// Skewness returns the skewness of the distribution.
+//
+// Always 0 for Logistic distribution
```
Comment (kortschak) [inline stat/distuv/logistic.go:79]: Dot.

File: stat/distuv/logistic_test.go [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+	want = l.Mu
+	if result := l.Mode(); result != want {
+		t.Errorf("Wrong mode value: %f != %f", result, want)
+	}
+}
+
+func TestLogisticStdDev(t *testing.T) {
+	t.Parallel()
+
+	const sq3 = 1.732050807568877293527446341505872366942805253810380 // sqrt(3)
```
Comment (kortschak) [inline stat/distuv/logistic_test.go:55]: If we have this constant here, perhaps move it into the package proper as `sqrt3` and then define `StdDev` as `return l.S * (math.Pi/sqrt3)`.

File: stat/distuv/logistic_test.go [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+	want = 1.0
+	if result := l.Variance(); result != want {
+		t.Errorf("Wrong Variance with Mu=%f, S=%f: %f != %f", l.Mu, l.S, result, want)
+	}
+}
+
+func TestLogisticCDF(t *testing.T) {
+	t.Parallel()
+
+	// values for "want" are taken from WolframAlpha: CDF[LogisticDistribution[mu,s], input] to 10 digits
```
Comment (kortschak) [inline stat/distuv/logistic_test.go:73]: Please lead with a capital and finish with a full stop.

File: stat/distuv/logistic_test.go [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+	input, want := 0.0, 0.5
+	if result := l.Survival(input); result != want {
+		t.Errorf("Wrong Survival(%f) with Mu=%f, S=%f: %f != %f", input, l.Mu, l.S, result, want)
+	}
+}
+
+func TestLogisticProb(t *testing.T) {
+	t.Parallel()
+
+	// values for "want" are taken from WolframAlpha: PDF[LogisticDistribution[mu,s], input] to 10 digits
```
Comment (kortschak) [inline stat/distuv/logistic_test.go:112]: Same here.

**Review by kortschak — CHANGES_REQUESTED** [review]
Some minor changes.

File: stat/distuv/logistic.go (hunk 6 above)
Comment (gudvinr) [reply stat/distuv/logistic.go:18]: Where do you want that? I deliberately skipped dot after link because dot is valid symbol in URL path and thus can be considered as part of URL so link will be invalid

Comment (kortschak) [reply stat/distuv/logistic.go:18]: The pkg.go.dev site is smart enough to not include trailing dots (for example, see https://pkg.go.dev/cmd/go#hdr-Calling_between_Go_and_C and the link to swig.org there). Valid concern though.

**Comment by gudvinr** [comment]
@kortschak can you review requested changes? I also rebased from new master.

**Review by kortschak — APPROVED** [review]
Thank you
